tools/perf/ucc_pt_coll_reduce.cc

```diff
@@ -0,0 +1,77 @@
+#include "ucc_pt_coll.h"
+#include "ucc_perftest.h"
+#include <ucc/api/ucc.h>
+#include <utils/ucc_math.h>
+#include <utils/ucc_coll_utils.h>
+
+ucc_pt_coll_reduce::ucc_pt_coll_reduce(ucc_datatype_t dt, ucc_memory_type mt,
+                        ucc_reduction_op_t op, bool is_inplace, bool is_root)
+{
+    has_inplace_   = true;
+    has_reduction_ = true;
+    has_range_     = true;
+    has_bw_        = true;
+    is_root_       = is_root;
+
+    coll_args.coll_type = UCC_COLL_TYPE_REDUCE;
+    coll_args.mask = 0;
+    if (is_inplace) {
+        coll_args.mask = UCC_COLL_ARGS_FIELD_FLAGS;
+        coll_args.flags = UCC_COLL_ARGS_FLAG_IN_PLACE;
+    }
+
+    coll_args.mask |= UCC_COLL_ARGS_FIELD_PREDEFINED_REDUCTIONS;
+    coll_args.reduce.predefined_op = op;
+    coll_args.root = 0;
+    coll_args.src.info.datatype = dt;
+    coll_args.src.info.mem_type = mt;
+    coll_args.dst.info.mem_type = mt;
+}
+
+ucc_status_t ucc_pt_coll_reduce::init_coll_args(size_t count,
+                                                   ucc_coll_args_t &args)
+{
+    size_t dt_size = ucc_dt_size(coll_args.src.info.datatype);
+    size_t size    = count * dt_size;
+    ucc_status_t st_src, st_dst;
+
+    args = coll_args;
+    args.src.info.count = count;
+    if (is_root_) {
+        UCCCHECK_GOTO(ucc_mc_alloc(&dst_header, size, args.dst.info.mem_type),
+                      exit, st_dst);
+        args.dst.info.buffer = dst_header->addr;
+    }
+    if (!is_root_ || !UCC_IS_INPLACE(args)) {
+        UCCCHECK_GOTO(ucc_mc_alloc(&src_header, size, args.src.info.mem_type),
+                      free_dst, st_src);
+        args.src.info.buffer = src_header->addr;
+    }
+    return UCC_OK;
+free_dst:
+    if (is_root_ && st_dst == UCC_OK) {
+        ucc_mc_free(dst_header);
+    }
+    return st_src;
+exit:
+    return st_dst;
+}
+
+void ucc_pt_coll_reduce::free_coll_args(ucc_coll_args_t &args)
+{
+    if (!is_root_ || !UCC_IS_INPLACE(args)) {
+        ucc_mc_free(src_header);
+    }
+    if (is_root_) {
+        ucc_mc_free(dst_header);
+    }
+}
+
+float ucc_pt_coll_reduce::get_bw(float time_ms, int grsize,
+                                    ucc_coll_args_t args)
+{
+    float S = args.src.info.count * ucc_dt_size(args.src.info.datatype);
```

Comment:
@bureddy why is this the way to calculate bw? 
for root itself, he will get way more than 1 vector along the algorithm (he will get ~ vector * radix* dist), and in general the amount of data in the network is larger

Reply:
This isn't implementation specific calculation. It is an estimation of best band width cascanario (for example ring implementation) using latency of our implementation (which is better latency) in order to analyze real bw of our implementation.